### PR TITLE
Remove storing of block and activeCollectionIDs from ResultManager

### PR DESCRIPTION
**Pull request by rajkharvar**
Old ResultManager gas consumption with same data: **362,913**
Tx: https://mumbai.polygonscan.com/tx/0x42825cc07c7ad747334d63e1a0fbb811116d6d6f4521b24dd95db922ca9b46ee

New ResultManager gas consumption with same data: **65,648**
https://mumbai.polygonscan.com/tx/0x5f532051a8febd123a13f070f0ce36a9a3f8679cdc6bcf4b2f894ff24abdfa85

* Update ResultManager contract
* Update Staking contract
* Update gas cost script

### CODE REVIEW

File: contracts/ResultManager.sol [hunk 1]
```diff
@@ -72,42 +67,35 @@ contract ResultManager is AccessControlEnumerable {
         );
 
         uint16[] memory ids = new uint16[](values.length);
```
Comment (SamAg19) [inline contracts/ResultManager.sol:69]: this can be removed

**Review by SamAg19 — CHANGES_REQUESTED** [review]

File: contracts/ResultManager.sol (hunk 1 above)
Comment (rajkharvar) [reply contracts/ResultManager.sol]: Resolved in [780bf9d](https://github.com/razor-network/bridge-proxy/pull/50/commits/780bf9de15de5fe8f5d208cddcf65a72dd79a60d)

File: contracts/ResultManager.sol [hunk 2]
```diff
@@ -72,42 +67,35 @@ contract ResultManager is AccessControlEnumerable {
...
-        blocks[epoch] = messageBlock;
         for (uint256 i; i < values.length; i++) {
             _collectionResults[values[i].collectionId] = values[i];
             collectionIds[values[i].name] = values[i].collectionId;
             ids[i] = values[i].collectionId;
```
[the remark this answers is not in the document]
Comment (rajkharvar) [reply contracts/ResultManager.sol]: Resolved in [780bf9d](https://github.com/razor-network/bridge-proxy/pull/50/commits/780bf9de15de5fe8f5d208cddcf65a72dd79a60d)

**Review by SamAg19 — APPROVED** [review]

**Review by SkandaBhat — DISMISSED** [review]

**Review by yohanelly95 — DISMISSED** [review]

File: test/Forwarder.js [hunk 3]
```diff
@@ -267,9 +267,11 @@ describe("Forwarder tests", () => {
     });
 
     it("Client shouldn't be allowed to getResult if stake is less than minStake", async () => {
```
Comment (SamAg19) [inline test/Forwarder.js:269]: Add a test case for calling to get result when above minStake and how changing the minStake param changes whitelist logic

**Review by SamAg19 — CHANGES_REQUESTED** [review]

File: contracts/Staking.sol [hunk 4]
```diff
@@ -89,28 +84,47 @@ contract Staking is AccessControlEnumerable {
...
+        // * remove the client address from stakersClients if amount staked == 0
+        if (stakersStakePerClient[msg.sender][client] == 0) {
+            uint length = stakersClients[msg.sender].length;
+
+            for (uint i = 0; i < length; i++) {
+                if (stakersClients[msg.sender][i] == client) {
+                    stakersClients[msg.sender][i] = stakersClients[msg.sender][
+                        length - 1
+                    ];
+                    stakersClients[msg.sender].pop();
```
[the remark this answers is not in the document]
Comment (rajkharvar) [reply contracts/Staking.sol:111]: Resolved in [223b3ba](https://github.com/razor-network/bridge-proxy/pull/50/commits/223b3bae899e0fac17a245c8e8c103799456f9dd)

File: test/Forwarder.js (hunk 3 above)
Comment (rajkharvar) [reply test/Forwarder.js:269]: Resolved in [223b3ba](https://github.com/razor-network/bridge-proxy/pull/50/commits/223b3bae899e0fac17a245c8e8c103799456f9dd)

**Review by SamAg19 — APPROVED** [review]

File: contracts/Staking.sol [hunk 5]
```diff
@@ -39,17 +58,79 @@ contract Staking is AccessControlEnumerable {
...
         uint256 amount = address(this).balance;
         payable(msg.sender).transfer(amount);
     }
+
+    function withdrawTokens() external onlyRole(ESCAPE_HATCH_ROLE) {
+        uint256 balance = token.balanceOf(address(this));
+        token.transfer(msg.sender, balance);
+    }
+
+    function stake(address client, uint256 amount) public {
```
Comment (SamAg19) [inline contracts/Staking.sol:77]: change from public to external

File: contracts/Staking.sol [hunk 6]
```diff
@@ -39,17 +58,79 @@ contract Staking is AccessControlEnumerable {
...
+        );
+        emit Staked(
+            msg.sender,
+            client,
+            stakersStakePerClient[msg.sender][client],
+            clientStake[client]
+        );
+    }
+
+    function unstake(address client, uint256 amount, uint256 index) public {
```
Comment (SamAg19) [inline contracts/Staking.sol:98]: change from public to external

**Review by SamAg19 — CHANGES_REQUESTED** [review]

File: contracts/Staking.sol (hunk 5 above)
Comment (rajkharvar) [reply contracts/Staking.sol]: Resolved in [14ade64](https://github.com/razor-network/bridge-proxy/pull/50/commits/14ade642a4d60f3afeb9030f384980f605bfdc90)

**Review by SamAg19 — APPROVED** [review]

**Review by SamAg19 — DISMISSED** [review]

File: contracts/Staking.sol [hunk 7]
```diff
@@ -39,17 +65,107 @@ contract Staking is AccessControlEnumerable {
...
+                token.transfer(_address, token.balanceOf(address(this))),
+                "token transfer failed"
+            );
+        } else {
+            revert("escape hatch is disabled");
+        }
+    }
+
+    function stake(address client, uint256 amount) external whenNotPaused {
+        require(amount > 0, "amount must be greater than 0");
```
Comment (SamAg19) [inline contracts/Staking.sol:105]: let's set a `minsafestake` which can be less than minstake. Here we are enabling any user to stake as long as they have even 1 wei of razor, which may allow few attack vectors in the future

**Review by SamAg19 — CHANGES_REQUESTED** [review]

**Review by yohanelly95 — DISMISSED** [review]

**Review by ashish10677 — DISMISSED** [review]

**Review by SamAg19 — APPROVED** [review]

**Review by yohanelly95 — APPROVED** [review]

**Pull request merged**